lib/teracy-dev-core/providers/manager.rb

```diff
@@ -0,0 +1,26 @@
+require 'teracy-dev/logging'
+
+module TeracyDevCore
+  module Providers
+    class Manager
+
+      @@logger = TeracyDev::Logging.logger_for(self)
+      @registry = {}
+
+      class << self
+        attr_reader :registry
+      end
+
+      def self.register(provider_type, provider_impl)
+        if !provider_impl.respond_to?(:configure)
+          @logger.warn("provider #{provider_impl} must implement configure method, ignored")
+          return
+        end
+
+        @registry[provider_type] = [provider_impl] if @registry[provider_type].nil?
+        @registry[provider_type] << provider_impl
```

Comment:
bug here:

```
[TeracyDevCore::Providers::Manager][DEBUG]: registry: {"virtualbox"=>[#<TeracyDevCore::Providers::VirtualBox:0x00000001020eaf80 @logger=#<Logger:0x00000001020eaf30 @level=0, @progname="TeracyDevCore::Providers::VirtualBox", @default_formatter=#<Logger::Formatter:0x00000001020eadf0 @datetime_format=nil>, @formatter=#<Proc:0x00000001020eab98@/Users/hoatle/teracy-dev/lib/teracy-dev/logging.rb:44>, @logdev=#<Logger::LogDevice:0x00000001020ead28 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x00000001020eacd8>>>>, #<TeracyDevCore::Providers::VirtualBox:0x00000001020eaf80 @logger=#<Logger:0x00000001020eaf30 @level=0, @progname="TeracyDevCore::Providers::VirtualBox", @default_formatter=#<Logger::Formatter:0x00000001020eadf0 @datetime_format=nil>, @formatter=#<Proc:0x00000001020eab98@/Users/hoatle/teracy-dev/lib/teracy-dev/logging.rb:44>, @logdev=#<Logger::LogDevice:0x00000001020ead28 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x00000001020eacd8>>>>], "vmware_desktop"=>[#<TeracyDevCore::Providers::VMware:0x0000000101266798 @logger=#<Logger:0x0000000101267440 @level=0, @progname="TeracyDevCore::Providers::VMware", @default_formatter=#<Logger::Formatter:0x00000001008578b0 @datetime_format=nil>, @formatter=#<Proc:0x00000001008541b0@/Users/hoatle/teracy-dev/lib/teracy-dev/logging.rb:44>, @logdev=#<Logger::LogDevice:0x00000001008571a8 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x0000000100856f50>>>>, #<TeracyDevCore::Providers::VMware:0x0000000101266798 @logger=#<Logger:0x0000000101267440 @level=0, @progname="TeracyDevCore::Providers::VMware", @default_formatter=#<Logger::Formatter:0x00000001008578b0 @datetime_format=nil>, @formatter=#<Proc:0x00000001008541b0@/Users/hoatle/teracy-dev/lib/teracy-dev/logging.rb:44>, @logdev=#<Logger::LogDevice:0x00000001008571a8 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x0000000100856f50>>>>]}
/Users/hoatle/teracy-dev/workspace/teracy-dev-core/lib/teracy-dev-core/providers/manager.rb:23:in `register'
```

Reply:
one provider_impl was registered into the registry twice, this is a bug

Reply:
logging with debug should help, try to use the debug log where we should use to check the changes.